lib/compress/zstdmt_compress.c

```diff
@@ -993,134 +1001,162 @@ size_t ZSTDMT_initCStream(ZSTDMT_CCtx* zcs, int compressionLevel) {
+            DEBUGLOG(4, "ZSTDMT_flushProduced: writing checksum : %08X \n", checksum);
+            MEM_writeLE32((char*)job.dstBuff.start + job.cSize, checksum);
+            job.cSize += 4;
+            zcs->jobs[wJobID].cSize += 4;
+            zcs->jobs[wJobID].frameChecksumNeeded = 0;
         }
-        {   size_t const toWrite = MIN(job.cSize - job.dstFlushed, output->size - output->pos);
-            DEBUGLOG(5, "ZSTDMT_flushNextJob: Flushing %u bytes from job %u ", (U32)toWrite, zcs->doneJobID);
+        assert(job.cSize >= job.dstFlushed);
+        if (job.dstBuff.start != NULL) {  /* one buffer present : some job is ongoing */
+            size_t const toWrite = MIN(job.cSize - job.dstFlushed, output->size - output->pos);
+            DEBUGLOG(5, "ZSTDMT_flushProduced: Flushing %u bytes from job %u (completion:%.1f%%)",
+                        (U32)toWrite, zcs->doneJobID, (double)job.consumed / job.srcSize * 100);
             memcpy((char*)output->dst + output->pos, (const char*)job.dstBuff.start + job.dstFlushed, toWrite);
             output->pos += toWrite;
             job.dstFlushed += toWrite;
-        }
-        if (job.dstFlushed == job.cSize) {   /* output buffer fully flushed => move to next one */
-            ZSTDMT_releaseBuffer(zcs->bufPool, job.dstBuff);
-            zcs->jobs[wJobID].dstBuff = g_nullBuffer;
-            zcs->jobs[wJobID].jobCompleted = 0;
-            zcs->doneJobID++;
-            zcs->consumed += job.srcSize;
-            zcs->produced += job.cSize;
-        } else {
-            zcs->jobs[wJobID].dstFlushed = job.dstFlushed;
-        }
-        /* return value : how many bytes left in buffer ; fake it to 1 if unknown but >0 */
+
+            if ( job.jobCompleted
+              && (job.dstFlushed == job.cSize) ) {   /* output buffer fully flushed => move to next one */
+                DEBUGLOG(5, "Job %u completed (%u bytes), moving to next one",
+                        zcs->doneJobID, (U32)job.dstFlushed);
+                ZSTDMT_releaseBuffer(zcs->bufPool, job.dstBuff);
+                zcs->jobs[wJobID].dstBuff = g_nullBuffer;
+                zcs->jobs[wJobID].jobCompleted = 0;
+                zcs->consumed += job.srcSize;
+                zcs->produced += job.cSize;
+                zcs->doneJobID++;
+            } else {
+                zcs->jobs[wJobID].dstFlushed = job.dstFlushed;   /* remember how much was flushed for next attempt */
+        }   }
+
+        /* return value : how many bytes left in buffer ; fake it to 1 when unknown but >0 */
         if (job.cSize > job.dstFlushed) return (job.cSize - job.dstFlushed);
-        if (zcs->doneJobID < zcs->nextJobID) return 1;   /* still some buffer to flush */
-        zcs->allJobsCompleted = zcs->frameEnded;   /* frame completed and entirely flushed */
-        return 0;   /* everything flushed */
-}   }
+        if (job.srcSize > job.consumed) return 1;   /* current job not completely compressed */
```

Comment:
Do we need `if (!job.jobCompleted) return 1;`? Otherwise, we could have consumed all the input, read all the output, but the job still has to produce one more chunk of output. Without it, setting `allJobsCompleted = zcs->frameEnded;` and returning `0` could be wrong.

Reply:
I've considered removing `job.jobCompleted` because it's (currently) the same signal as `job.srcSize == job.consumed`.

The worker thread uses the synchronous bufferless API, so there is no "data retention" inside context : any data read immediately produces a compressed block. When `job.srcSize == job.consumed`, we have reached the end of job, and produced all the compressed content.

Reply:
Ahh, consumed is only updated after the compression is completed. Can you add either add an assert, or go ahead and remove `jobCompleted`?

Reply:
Removing `jobCompleted` introduces a new problem : 
in some (rare) cases, it is necessary to create a last 0-size job just to finish the frame.
In which case, `job.srcSize == 0 == job.consumed` immediately, even though the job itself (creating an empty block with "end of frame" marker) is not yet done.

Potential mitigations : 
1) keep `jobCompleted` signal
2) add some special code to create a last 0-size block without going through the job creation process.
    + sounds like an obvious optimisation.
        However, it adds complexity, for a very rare use case. 
        More importantly, `zstdmt` is up to now agnostic of the frame format. Creating such a `0-size` last block directly inside `zstdmt_compress.c` makes the frame format a new dependency of this module. Alternatively, the `0-size` last block could be done inside `zstd_compress.c` and invoked from `zstdmt_compress.c`, though this is a very specialised entry point without another user for the time being.

Reply:
finally decided to remove `job.jobCompleted`
and use `job.consumed==job.srcSize` instead,
complemented with `ZSTD_writeLastEmptyBlock()` for the special case `job.srcSize==0`.